libsolidity/ast/Types.h

```diff
@@ -372,6 +373,14 @@ class Type
 	/// are returned without modification.
 	virtual TypeResult interfaceType(bool /*_inLibrary*/) const { return nullptr; }
 
+	/// @returns list of subtypes for composite types (base type for array, members for structs,
+	/// value type for mappings) or empty list for ordinary types
+	virtual std::vector<Type const*> subtypes() const { return {}; }
```

Comment:
I thought `subtypes()` needs specialized implementation for `MappingType`, `StructType`, `ArrayType`, `ArraySliceType`. I thought of introducing `CompositeType`, and deriving `MappingType` and `ReferenceType` from it, but was not sure if it is justified.

I do see your point regarding `sizeUpperBound() = 0`, but afraid making the PR too invasive. Could we come up with some reasonable conservative default value?

Reply:
Your default makes sense for simple types so if your goal is to avoid changing type classes other than those you listed then I don't think we can come up with a better one.

There are some other functions in `Type` that have defaults like this so I think it would also be fine to leave it as is in this PR and refactor all of them in a separate one. But really, it's not super important since the code is not broken, just wanted to point out that this code would have a lower chance of silently breaking in the future without a default in my opinion.

Reply:
> Even without introducing an intermediate class, the change would require just adding a one-line function in each of the type classes.

Sure. I meant to introduce `CompositeType` with `subtypes() = 0;` so only the derived classes would need to define the function. And we would not recurse into non-composite types at all. As a side effect, `CompositeType` would allow to get rid of the ugly `if (referenceType || mappingType)`.

But as you said, the issue might be not that big :).

Reply:
`CompositeType` does sound like a good idea to me. Maybe a common base class for simple types would make sense too? If you just don't want to add it in this PR you could still create a separate refactoring PR and get it merged first. Then rebase this on it.